pkg/controller/attach_detach.go

```diff
@@ -110,7 +110,18 @@ func (r *ReconcileAttachDetach) Reconcile(ctx context.Context, request reconcile
 
 	// deletion request
 	if objectDeletionRequested(azVolumeAttachment) {
+		r.SharedState.nodeDiskAvailabilityMapLock.Lock()
```

Comment:
You can avoid the need for a global lock by using [`*atomic.Int32`](https://pkg.go.dev/sync/atomic#Int32) as the value in the `nodeDiskAvailabilityMap`. You need to store a pointer to the `atomic.Int32` in the `sync.Map` since you need to call methods on a shared value in order to update the count.

You can use [`atomic.Int32.Add(1)`](https://pkg.go.dev/sync/atomic@go1.19.3#Int32.Add) to increment the count. Decrementing the count is more complex since you need to use [`atomic.Int32.Load()`](https://pkg.go.dev/sync/atomic@go1.19.3#Int32.Load) to get the current value and then [`atomic.Int32.CompareAndSwap(oldValue, newValue int)`](https://pkg.go.dev/sync/atomic@go1.19.3#Int32.CompareAndSwap) to store the updated value if the current value is not zero. This needs to be done in a loop since another go routine may race to update the value, e.g.

```go
currentCapacity := 0
remainingCapacity, nodeExists := r.SharedState.nodeDiskAvailabilityMap.Load(azVolumeAttachment.Spec.NodeName)
if nodeExists && remainingCapacity != nil {
    for {
        currentCapacity = remainingCapacity.(*atomic.Int32).Load()
        if currentCapacity == 0 || remainingCapacity.(*atomic.Int32).CompareAndSwap(currentCapacity, currentCapacity - 1) {
            break
    }
}
```

Reply:
Done